libbeat/processors/add_cloud_metadata/provider_aws_ec2.go

```diff
@@ -18,125 +18,134 @@
 	if err != nil {
-		logger.Warnf("error when reading token request for getting IMDSv2 token: %s. No token in the metadata request will be used.", err)
-		return ""
+		logger.Warnf("error fetching cluster name metadata: %s.", err)
 	}
 
-	return string(all)
-}
+	accountID := instanceIdentity.InstanceIdentityDocument.AccountID
 
-// AWS EC2 Metadata Service
-var ec2MetadataFetcher = provider{
-	Name: "aws-ec2",
+	_, _ = result.metadata.Put("instance.id", instanceIdentity.InstanceIdentityDocument.InstanceID)
+	_, _ = result.metadata.Put("machine.type", instanceIdentity.InstanceIdentityDocument.InstanceType)
+	_, _ = result.metadata.Put("region", awsRegion)
+	_, _ = result.metadata.Put("availability_zone", instanceIdentity.InstanceIdentityDocument.AvailabilityZone)
+	_, _ = result.metadata.Put("account.id", accountID)
+	_, _ = result.metadata.Put("image.id", instanceIdentity.InstanceIdentityDocument.ImageID)
 
-	Local: true,
+	// for AWS cluster ID is used cluster ARN: arn:partition:service:region:account-id:resource-type/resource-id, example:
+	// arn:aws:eks:us-east-2:627286350134:cluster/cluster-name
+	if clusterName != "" {
+		clusterARN := fmt.Sprintf("arn:aws:eks:%s:%s:cluster/%v", awsRegion, accountID, clusterName)
 
-	Create: func(_ string, config *conf.C) (metadataFetcher, error) {
-		ec2Schema := func(m map[string]interface{}) mapstr.M {
-			m["serviceName"] = "EC2"
-			out, _ := s.Schema{
-				"instance":          s.Object{"id": c.Str("instanceId")},
-				"machine":           s.Object{"type": c.Str("instanceType")},
-				"region":            c.Str("region"),
-				"availability_zone": c.Str("availabilityZone"),
-				"service": s.Object{
-					"name": c.Str("serviceName"),
-				},
-				"account": s.Object{"id": c.Str("accountId")},
-				"image":   s.Object{"id": c.Str("imageId")},
-			}.Apply(m)
-			return mapstr.M{"cloud": out}
-		}
+		_, _ = result.metadata.Put("orchestrator.cluster.name", clusterName)
+		_, _ = result.metadata.Put("orchestrator.cluster.id", clusterARN)
+	}
+}
 
-		headers := make(map[string]string, 1)
-		token := getIMDSv2Token(config)
-		if len(token) > 0 {
-			headers[ec2InstanceIMDSv2TokenValueHeader] = token
-		}
+func fetchEC2ClusterNameTag(awsConfig awssdk.Config, instanceID string) (string, error) {
+	svc := NewEC2Client(awsConfig)
+	input := &ec2.DescribeTagsInput{
+		Filters: []types.Filter{
+			{
+				Name: awssdk.String("resource-id"),
+				Values: []string{
+					instanceID,
+				},
+			},
+			{
+				Name: awssdk.String("key"),
+				Values: []string{
+					"eks:cluster-name",
+				},
+			},
+		},
+	}
 
-		fetcher, err := newMetadataFetcher(config, "aws", headers, metadataHost, ec2Schema, ec2InstanceIdentityURI)
-		return fetcher, err
-	},
+	tagsResult, err := svc.DescribeTags(context.TODO(), input)
+	if err != nil {
+		return "", fmt.Errorf("error fetching EC2 Tags: %w", err)
+	}
+	if len(tagsResult.Tags) > 0 {
```

Comment:
quick question here: since we are only returning `Tags[0]`, does that mean we only have one set of tag key/value returned from the DescribeTags API call? If so, should we check len(tagsResult.Tags) == 1 instead?

Reply:
yes, there only 1 element is expected, it is basically the same as to run:
```
instance_id=$(curl http://169.254.169.254/latest/meta-data/instance-id)
aws ec2 describe-tags --filter Name=resource-id,Values=$instance_id Name=key,Values=eks:cluster-name --region us-east-2
{
    "Tags": [
        {
            "Key": "eks:cluster-name",
            "ResourceId": "i-020c696ca32297693",
            "ResourceType": "instance",
            "Value": "cluster-name"
        }
    ]
}
```

Reply:
@kaiyan-sheng I've changed it in https://github.com/elastic/beats/pull/35182/commits/79070e6ec78a77b72227fc0b7877934227beeccd
I've found a mistake I make in `types.TagDescription` struct, now it return exactly 1 element